src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php

```diff
@@ -91,9 +91,18 @@ protected function doLoadChoicesForValues(array $values, ?callable $value): arra
             trigger_deprecation('symfony/doctrine-bridge', '5.1', 'Not defining explicitly the IdReader as value callback when query can be optimized is deprecated. Don\'t pass the IdReader to "%s" or define the "choice_value" option instead.', __CLASS__);
         }
 
+        $idReader = null;
+        if (\is_array($value) && $value[0] instanceof IdReader) {
+            $idReader = $value[0];
+        } elseif ($value instanceof \Closure && ($rThis = (new \ReflectionFunction($value))->getClosureThis()) instanceof IdReader) {
+            $idReader = $rThis;
+        } elseif ($legacy) {
+            $idReader = $this->idReader;
+        }
```

Comment:
All right, this was a tough one. 😓 

`DoctrineChoiceLoader` inspects the callable `$value` to check if it's an array where it can extract an `IdReader` from. Now that we wrap that callable in a closure, that check failed silently which broke the optimization code here and resulted in all entities being loaded.

I "fixed" that by extracting the `IdReader` from the closure now, but I really believe `DoctrineChoiceLoader` should not make that kind of assumption, really.

Reply:
Backported as #42135 to make the high deps tests green.

Reply:
well, using the IdReader directly when the default value callable is used is what allows us to optimize the loading of the selected choice (as we know that the value is the DB id in that case)

Reply:
I understand that. I'm not questioning the "what", merely the "how"…

Reply:
Well, the condition for applying the optimization is "the value callback returns the id of the entity". Given that the `choice_value` option is configuration by users, we have no other choice than inspecting it.